### PR TITLE
Added bsg_print_stats to cuda lite runtime regression tests

### PR DESCRIPTION
**Pull request by bornaehsani**
- Added a new symbol, __bsg_tile_group_id which is the flat tile group ID, or (__bsg_tile_group_id_y * __bsg_grid_dim_x + __bsg_tile_group_id_x) to the bsg_set_tile_x_y().
- Added bsg_print_stat() statements to the beginning and end of all cuda lite runtime regression tests.

### CODE REVIEW

File: software/bsg_manycore_lib/bsg_set_tile_x_y.h [hunk 1]
```diff
@@ -7,6 +7,7 @@ extern int __bsg_grid_dim_x;	  //The X Dimensions of the grid of tile groups
 extern int __bsg_grid_dim_y;	  //The Y Dimensions of the grid of tile groups
 extern int __bsg_tile_group_id_x; //The X Cord of the tile group within the grid
 extern int __bsg_tile_group_id_y; //The Y Cord of the tile group within the grid
+extern int __bsg_tile_group_id;   //The flat tile group id within the grid
```
Comment (tommydcjung) [inline software/bsg_manycore_lib/bsg_set_tile_x_y.h]: these definitions seem little confusing.. does the grid mean the entire manycore array?

Comment (drichmond) [reply software/bsg_manycore_lib/bsg_set_tile_x_y.h]: I agree with tommy. Is there a place that `grid` is defined? Perhaps it should be defined in this file.

File: software/spmd/bsg_cuda_lite_runtime/dram_device_allocated/kernel_dram_device_allocated.c [hunk 2]
```diff
@@ -23,5 +26,10 @@ int  __attribute__ ((noinline)) kernel_dram_device_allocated(int *addr, int bloc
 
 	bsg_tile_group_barrier(&r_barrier, &c_barrier); 
 
+	if (__bsg_id == 0)
+		bsg_print_stat(1000 + __bsg_tile_group_id);
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/dram_device_allocated/kernel_dram_device_allocated.c]: What's the purpose of 1000? It seems like a magic number

File: software/spmd/bsg_cuda_lite_runtime/dram_host_allocated/kernel_dram_host_allocated.c [hunk 3]
```diff
@@ -10,9 +10,17 @@ INIT_TILE_GROUP_BARRIER(r_barrier, c_barrier, 0, bsg_tiles_X-1, 0, bsg_tiles_Y-1
...
 
+	if (__bsg_id == 0)
+		bsg_print_stat(__bsg_tile_group_id);
+
 	*addr = 0x1234;
 
 	bsg_tile_group_barrier(&r_barrier, &c_barrier); 
 
+	if (__bsg_id == 0)
+		bsg_print_stat(1000 + __bsg_tile_group_id);
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/dram_host_allocated/kernel_dram_host_allocated.c]: DItto here.

File: software/spmd/bsg_cuda_lite_runtime/float_all_ops/kernel_float_all_ops.c [hunk 4]
```diff
@@ -41,5 +43,10 @@ int  __attribute__ ((noinline)) kernel_float_all_ops(float *A, float *B,
 
 	bsg_tile_group_barrier(&r_barrier, &c_barrier); 
 
+	if (__bsg_id == 0)
+		bsg_print_stat(1000 + __bsg_tile_group_id);
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/float_all_ops/kernel_float_all_ops.c]: And here

File: software/spmd/bsg_cuda_lite_runtime/float_matrix_mul/kernel_float_matrix_mul.c [hunk 5]
```diff
@@ -33,5 +35,10 @@ int  __attribute__ ((noinline)) kernel_float_matrix_mul(float *A, float *B, floa
 
 	bsg_tile_group_barrier(&r_barrier, &c_barrier); 
 
+	if (__bsg_id == 0)
+		bsg_print_stat(1000 + __bsg_tile_group_id);
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/float_matrix_mul/kernel_float_matrix_mul.c]: Here too

File: software/spmd/bsg_cuda_lite_runtime/float_matrix_mul_shared_mem/kernel_float_matrix_mul_shared_mem.c [hunk 6]
```diff
@@ -124,5 +127,10 @@ int  __attribute__ ((noinline)) kernel_float_matrix_mul_shared_mem(float *A, flo
 
 	bsg_tile_group_barrier (&r_barrier, &c_barrier) ;
 
+	if (__bsg_id == 0)
+		bsg_print_stat(1000 + __bsg_tile_group_id);
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/float_matrix_mul_shared_mem/kernel_float_matrix_mul_shared_mem.c]: Ditto.

File: software/spmd/bsg_cuda_lite_runtime/float_vec_add/kernel_float_vec_add.c [hunk 7]
```diff
@@ -12,12 +12,20 @@ INIT_TILE_GROUP_BARRIER(r_barrier, c_barrier, 0, bsg_tiles_X-1, 0, bsg_tiles_Y-1
...
+
 	int start_x = block_size_x * (__bsg_tile_group_id_y * __bsg_grid_dim_x + __bsg_tile_group_id_x); 
 	for (int iter_x = __bsg_id; iter_x < block_size_x; iter_x += bsg_tiles_X * bsg_tiles_Y) { 
 		C[start_x + iter_x] = A[start_x + iter_x] + B[start_x + iter_x];
 	}
 
 	bsg_tile_group_barrier(&r_barrier, &c_barrier); 
 
+	if (__bsg_id == 0)
+		bsg_print_stat(1000 + __bsg_tile_group_id);
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/float_vec_add/kernel_float_vec_add.c]: Y lo mismo aqui.

File: software/spmd/bsg_cuda_lite_runtime/float_vec_add_shared_mem/kernel_float_vec_add_shared_mem.c [hunk 8]
```diff
@@ -56,6 +59,10 @@ int  __attribute__ ((noinline)) kernel_float_vec_add_shared_mem(float *A, float
 
 	bsg_tile_group_barrier(&r_barrier, &c_barrier);
 
+	if (__bsg_id == 0)
+		bsg_print_stat(1000 + __bsg_tile_group_id);
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/float_vec_add_shared_mem/kernel_float_vec_add_shared_mem.c]: I think you get the point...

File: software/bsg_manycore_lib/bsg_set_tile_x_y.h (hunk 1 above)
Comment (bornaehsani) [reply software/bsg_manycore_lib/bsg_set_tile_x_y.h]: Good point, I should have explained more. The grid is actually a 2D structure or array of tile groups that represent the entire application, with each element being a separate tile group with X,Y coordinates inside the grid. The term came from CUDA, where a grid is a 3D array of thread blocks.
I added some comments on the top of the file, what do you think?

File: software/spmd/bsg_cuda_lite_runtime/dram_device_allocated/kernel_dram_device_allocated.c (hunk 2 above)
Comment (bornaehsani) [reply software/spmd/bsg_cuda_lite_runtime/dram_device_allocated/kernel_dram_device_allocated.c]: Oh it is a magic number, but I don't have a better way of doing things at the moment. 
You see, with each bsg_print_stats() instruction, a line of stats is printed out for every tile in the manycore, even with the "if (__bsg_id =0)" condition.
And I need two of these print_stats, one at the beginning and one at the end of kernel, so I can subtract the cycles and calculate the total execution time, and I need a way to distinct between the two statements using the tag, but the tag should also indicate which tile group the statement is coming from, since multiple tile groups might use the same tile one after another. 
So basically statements with tags <1000 are beginning statements, and those with tag >1000 are end statements. Do you have any ideas how I can do this without using a magic number? 
I guess I can use positive and negative tags, how does that sound?

File: software/spmd/bsg_cuda_lite_runtime/float_vec_add_shared_mem/kernel_float_vec_add_shared_mem.c (hunk 8 above)
Comment (bornaehsani) [reply software/spmd/bsg_cuda_lite_runtime/float_vec_add_shared_mem/kernel_float_vec_add_shared_mem.c]: si, entiendo el punto :D

File: software/spmd/bsg_cuda_lite_runtime/dram_device_allocated/kernel_dram_device_allocated.c (hunk 2 above)
Comment (bornaehsani) [reply software/spmd/bsg_cuda_lite_runtime/dram_device_allocated/kernel_dram_device_allocated.c]: Actually negative and positive wouldn't work for tile group 0. So I'm not sure if there is a better way to do this, do you have any ideas?

Comment (tommydcjung) [reply software/spmd/bsg_cuda_lite_runtime/dram_device_allocated/kernel_dram_device_allocated.c]: I think it's fine to pick whatever value for the print_stat_tag, as long as it's easily identifiable.

Comment (drichmond) [reply software/spmd/bsg_cuda_lite_runtime/dram_device_allocated/kernel_dram_device_allocated.c]: I don't mind the number, I would rather have it be defined as a macro. 

And I think it would be better represented as a hex number, so that the binary interpretation (if ever needed) is easily distinguishable

Comment (tommydcjung) [reply software/spmd/bsg_cuda_lite_runtime/dram_device_allocated/kernel_dram_device_allocated.c]: yup, macro would be better. I used 0xdead in the past.

File: software/bsg_manycore_lib/bsg_set_tile_x_y.h (hunk 1 above)
Comment (drichmond) [reply software/bsg_manycore_lib/bsg_set_tile_x_y.h]: Seems good to me. @tommydcjung?

File: software/spmd/bsg_cuda_lite_runtime/dram_device_allocated/kernel_dram_device_allocated.c [hunk 9]
```diff
@@ -23,5 +26,10 @@ int  __attribute__ ((noinline)) kernel_dram_device_allocated(int *addr, int bloc
 
 	bsg_tile_group_barrier(&r_barrier, &c_barrier); 
 
+	if (__bsg_id == 0)
```
Comment (drichmond) [inline software/spmd/bsg_cuda_lite_runtime/dram_device_allocated/kernel_dram_device_allocated.c]: How is this different than simply checking x==0 && y == 0?

File: software/bsg_manycore_lib/bsg_set_tile_x_y.h [hunk 10]
```diff
@@ -1,3 +1,21 @@
+/* Each application launched onto the manycore is represented by a grid of tile groups.        */
+/* A grid is a two diemsnional structure represeting the entire application, with each         */
+/* element of grid being a tile group with X and Y coordinates bsg_tile_group_id_x/y,          */
+/* and the flat coordiante of bsg_tile_group_id, which is calculate using grid dimensions      */
+/* __bsg_x               --> The X coordinate of tile inside its tile group                    */
+/* __bsg_y               --> The Y coordiante of tile inside its tile group                    */
+/* __bsg_id              --> The flat index of tile inside tile group                          */
+/*                            __bsg_id = __bsg_y * __bsg_tiles_X + __bsg_x                     */
```
Comment (tommydcjung) [inline software/bsg_manycore_lib/bsg_set_tile_x_y.h]: would be nice to add that __bsg_tiles_X is the x-dimension of the tile-group. also for __bsg_tiles_Y.

File: software/bsg_manycore_lib/bsg_set_tile_x_y.h [hunk 11]
```diff
@@ -1,3 +1,21 @@
...
+/* __bsg_grp_org_x       --> The X coordinate of the tile group origin tile                    */
+/* __bsg_grp_org_y       --> The Y coordinate of the tile group origin tile                    */
+/* __bsg_grid_dim_x      --> X dimension of grid, or number of tile groups in the X dimensions */
```
Comment (tommydcjung) [inline software/bsg_manycore_lib/bsg_set_tile_x_y.h]: are you saying that "grid" is 2D array of "tile groups"?
and a tile group is a 2D array of tiles?

File: software/bsg_manycore_lib/bsg_set_tile_x_y.h [hunk 12]
```diff
@@ -1,3 +1,21 @@
...
+/* __bsg_grid_dim_y      --> Y dimension of grid, or number of tile groups in the Y dimensions */
+/* __bsg_tile_group_id_x --> X coordinate of tile group within the application's grid          */
+/* __bsg_tile_group_id_y --> Y coordinate of tile group within the application's grid          */
+/* __bsg_tile_group_id_x --> flat index of tile group within the application's grid            */
```
Comment (tommydcjung) [inline software/bsg_manycore_lib/bsg_set_tile_x_y.h]: did you mean to say "__bsg_tile_group_id"? for line 15

**Comment by drichmond** [comment]
Where are we on this PR?

**Comment by bornaehsani** [comment]
Stale, plan was to replace all instances of print_stat in kernels with one print_stat inside cuda lite runtime, which ran into a dead-end, and I haven't had time to solve that since. but I will close this PR for now and make an issue, and make a new PR when it's fixed.
